### ISSUE TITLE
MXE build of Modelcompiler not working on Windows

### ISSUE THREAD

**Issue by impaktor** [pioneerspacesim/pioneer#3449]
Just opening this to keep track of the, otherwise hard to find, discussion over in:
https://github.com/pioneerspacesim/pioneer/commit/f4c4ddd6cbe1d8735edfc6cb505197d7e88eeab8

**Comment by fluffyfreak** [comment]
Changed the title of this since it's the `modelcompiler` that processes models into our SGM format for faster loading, and not the built-in `pioneer -mv` option for viewing models, that is broken.

**Comment by fluffyfreak** [comment]
@robn I think this is the cause, or at least it's a problem, to do with the recent builds not working, the `modelcompiler` can't rebuild the `sgm` files, so it ships the **old** `sgm` files with the new build.

Those old sgm files cause the game to crash.

**Comment by robn** [comment]
Presumably I screwed something when I crashed us over to Ubuntu build hosts the other week. Will try to look tomorrow.

**Comment by impaktor** [comment]
@robn is it possible to revert the download link on the homepage to a non-crashing version in the mean time?

**Comment by fluffyfreak** [comment]
I can verify this with my own MXE build of `modelcompiler` but not the why, It's very strange.

Pioneer works just fine.

**Comment by fluffyfreak** [comment]
AHA!!! @robn it is running, sort of, but then something it obviously silently failing.

I've had to fall back on a sort of `printf` debugging by writing out txt files when it gets to certain stage in the program which is... primitive to say the least.

...and it all seems to be working actually, it doesn't output anything to the cmd line like it should but with the `-b inplace` parameters it does produce the `sgm` files in the `data/models/` sub-folders. It just does it so fast (`on my ssd`) that I didn't even notice it run, sub 1-second compile time!!!

Anyway Pioneer loads just fine with those generated files now.

In short, I do not know why the build machine is failing to produce valid sgm files :(

**Comment by fluffyfreak** [comment]
Also (@impaktor, @robn) I have a zipped version that I could upload with the fixed `sgm` files and my `modelcompiler.exe` that seems to work. Should I replace the current bad `pioneer-20151109-win32.7z` version with a `pioneer-20151109-win32-fixed.7z` upload?

**Comment by robn** [comment]
Yes, just upload it straight over the top.

I will figure out why the build isn't generating correct `sgm` files before I do the next build. I hope to have time on Friday.

**Comment by fluffyfreak** [comment]
:D

**Comment by impaktor** [comment]
> Should I replace the current bad pioneer-20151109-win32.7z version with a pioneer-20151109-win32-fixed.7z upload?

Hell yes! You have access?

**Comment by fluffyfreak** [comment]
@impaktor yeah I have access now that I've recovered my SF account :)
I've uploaded a direct replacement of the file so it's go the same name and has replaced the existing win32 download.

**Comment by robn** [comment]
Sorry for taking so long!

```
<robn> so what happens is that the build script essentially does for $platform in linux64 linux32 win32 { build; archive }
<robn> part of archive is "collect all the files"
<robn> and for non-linux64 builds, that includes "grab .sgm files from linux64 build"
<robn> because we need to run modelviewer to generate them at build time, only for that platform because that what the build host can run
<robn> and that works great
<robn> except
<robn> the list of platforms to build come from the keys of a hash
<robn> and they get randomised between runs
<robn> so if we end up doing, say
<robn> for $platform in linux32 win32 linux64
<robn> then the archive step for linux32 and win32 will copy the sgm files from the previous run
```

So I think that's fixed; please try these before I do a proper build:

http://pioneerspacesim.net/temp/pioneer-20151203-win32.7z
http://pioneerspacesim.net/temp/pioneer-20151203-linux64.tar.bz2
http://pioneerspacesim.net/temp/pioneer-20151203-linux32.tar.bz2

**Comment by Brianetta** [comment]
Work-around: Build it twice. (-:

**Comment by robn** [comment]
Quite!

**Comment by robn** [comment]
Ok, 20151204 is uploaded and will appear on the homepage any moment now, so I think my part of this is fixed. I don't know if it addresses the original issue so I'll leave the issue open.

**Comment by impaktor** [comment]
@robn in #3527 it appears that one of the builds from last month did not sync with transifex, leading to missing strings, causing crash when not using English (in Options sceen due to AUTO_SAVE string missing). If that was a one time fluke then I guess we can close it, if not, it would be good if we understood why the build failed to communicate with transifex.

**Comment by fluffyfreak** [comment]
Something about this only just occurred to me and that is: __Shouldn't the SGM file be binary identical for all platforms anyway?__

I think this needs some looking into further.

**Comment by walterar** [comment]
Long ago I'm using (and publishing) x.sgm files generated in Linux 64, in common /data to Linux 64/32 and Windows 32 (MXE) No problems.

**Comment by fluffyfreak** [comment]
Just updating this to say that I think the MXE build is still broken in the latest download.

**Comment by walterar** [comment]
With Wine on Ubuntu works well.
With Windows 7, simply does nothing.

**Comment by fluffyfreak** [comment]
thanks @walterar that's interesting, very strange but interesting!

**Comment by impaktor** [comment]
Link #4446

**Comment by The-EG** [comment]
This isn't really applicable any longer, as it seems we aren't using MXE and we are now using cmake instead of autotools as well.

In addition, I'll be switching modelcompiler back to being a console app in #4982 , and I've verified the installer built there in appveyor works, including generating SGMs.
